### PR TITLE
[Build] Generate Bazel config for compiling with clang and libc++ in CI

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
Currently when we want to compile the project with `clang`, we specify `CC=clang` which still uses `libstdc++` from GCC toolchain. There are compile failures when upgrading C++ standard to 17. This PR generates a config to use `clang` together with `libc++` when Bazel config `--config=llvm` is specified. The config generation can be ran from CI or other Ubuntu systems. Developers on MacOS don't need this config because this toolchain is already the default.

Rolling forward the upgrade of C++ standard to 17 from Yi (#18455), which seems to work now.
The C++ standard upgrade requires running C++ / ASAN / TSAN tests with `--config=llvm` as well. With `libc++` ASAN enabled `container-overflow` check which uncovers two issues that are fixed.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mwtian** [comment]
Windows failure seems unrelated.

**Comment by scv119** [comment]
this looks great! 
for my understanding; with this fix, we don't need clang for building C++17; but only when we want to run tests with clang (asan/tsan) as well as clang-tidy, we need to run the install-clang script.

File: ci/travis/install-llvm-binaries.sh [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    linux-gnu-ubuntu*)
+      printInfo "Downloading LLVM from ${url}"
+      wget -c $url -O llvm.tar.xz
+      printInfo "Installing LLVM to ${targetdir}"
+      mkdir -p "${targetdir}"
+      tar -xf ./llvm.tar.xz -C "${targetdir}" --strip-components=1
+      rm llvm.tar.xz
+      ;;
+    *)
+      printError "Unknown linux distro ${OSTYPE}"
```
Comment (scv119) [inline ci/travis/install-llvm-binaries.sh:70]: maybe "Unsupported linux distro..."

**Review by scv119 — APPROVED** [review]

**Comment by scv119** [comment]
cc @iycheng maybe take a look?

**Review by fishbone — APPROVED** [review]
Really nice work!

**Comment by mwtian** [comment]
Windows test failures should be unrelated.

**Pull request merged**